internal/controller/apiextensions/composite/reconciler.go

```diff
@@ -608,6 +659,22 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 		if kerrors.IsConflict(err) {
 			return reconcile.Result{Requeue: true}, nil
 		}
+
+		meta := r.handleCommonCompositionResult(ctx, res, xr, log)
```

Comment:
This `handleCommonCompositionResult` method is the only part of the implementation that I'm not sure about. It's more intuition (or taste?) than anything concrete. I think part of what's bugging me is there isn't a very clean separation of responsibilities. For example `handleCommonCompositionResult` sets some conditions, but we also set some conditions here in `Reconcile`.

I don't think this needs to be a blocker though. I'd like to take a pass and see if I can refactor it a bit, but I'm happy to do that after this is merged if you're okay with it. My focus is temporarily on another project, so I don't want to delay this further waiting to get time to play with it. 😄

Reply:
Will do! I probably won't get time for a few weeks.